proposals/pattern-matching.md

```diff
@@ -0,0 +1,714 @@
+        _
+    ) -> // ...
+}
+```
+
+### Nested patterns using `is` <a name="use-nested-is"></a>
+
+This suggestion involves re-using the keyword `is` every time a nested pattern matched is performed. Potentially, it could allow simply recursing on `when` conditions in general:
+
+Consider the original proposal syntax:
+```kotlin
+val result = when(download) {
+  is App(name, Person("Alice", _)) -> "Alice's app $name"
+  is Movie(title, Person("Alice", _)) -> "Alice's movie $title"
+  is App, Movie -> "Not by Alice"
+}
+```
+... as opposed to the alternative:
+```kotlin
+val result = when(download) {
+  is App(name, is Person("Alice", in 0..18)) -> "Alice's app $name"
+  is Movie(val title, Person("Alice", _)) -> "Alice's movie $title"
+  is App, Movie -> "Not by Alice"
+}
+```
+The argument in favour of this syntax is that the added keywords clarify any
+ambiguities. This proposal argues **against** this more verbose syntax for
+the following reasons:
+ - It lifts the [literals only](#literals-only) restriction, which would
+ allow the user to inline things inside a pattern match (making for harder to
+ read equality checks)
+ - It is much more verbose to write for nested patterns of more than one
+ level, or where there is a lot of inspection of destructured elements.
+ Consider how using `is` and `val` every time would look for the 
+ [Arrow example](#arrow-example), which does not even make use of
+more than 1 level of nesting nor guards:
+  
+```kotlin
+fun <A> Kind<ForOption, A>.eqK(other: Kind<ForOption, A>, EQ: Eq<A>) =
+    when(this.fix() to other.fix()) {
+        is (is Some(val a), is Some(val b)) -> EQ.run { a.eqv(b) }
+        is (is None, is None) -> true
+        else -> false
+    }
+
+
+fun Ior<L, R>.eqv(other: Ior<L, R>): Boolean = when (this to other) {
+    is (is Ior.Left(val a), is Ior.Left(val b)) -> EQL().run { a.eqv(b) }
+    is (is Ior.Both(val al, val ar), is Ior.Both(val bl, val br)) ->
+      EQL().run { al.eqv(bl) } && EQR().run { ar.eqv(br) }
+    is (is Ior.Right(val a), is Ior.Right(val b)) -> EQR().run { a.eqv(b) }
+    else -> false
+}
+```
+
+#### Conclusion
+
+There are two competing alternative syntax possibilites for pattern matching
+in Kotlin. While this proposal argues for one that
+- is slightly more restrictive (no nested `when` conditions)
+- is less verbose
+- encourages guards
+
+...using `val` and `is` would make for a pattern mathcing that
+  - is more verbose (but offers some clarity)
+  - allows inlining for equality checking
+  - allows recursive `when` conditions (like `in 0..18`)
+
+### Restricting matching to data classes only
+
+A possibilty suggested during the conception of this proposal was to restrict pattern matching to data classes. The argument would be to start with a samller size of 'matchable' elements in order to keep the inital proposal and feature as simple as possible, as it could be extended later down the line.
```

Comment:
```suggestion
A possibility suggested during the conception of this proposal was to restrict pattern matching to data classes. The argument would be to start with a smaller size of 'matchable' elements in order to keep the initial proposal and feature as simple as possible, as it could be extended later down the line.
```